tests/rustc_tests.rs

```diff
@@ -0,0 +1,761 @@
+error: foo
+ --> test.rs:3:6
+  |
+3 |      X0 Y0 Z0
+  |  _______^
+4 | |    X1 Y1 Z1
+  | | _________^
+5 | ||   X2 Y2 Z2
+  | ||____^ `X` is a good letter
+6 |  |   X3 Y3 Z3
+  |  |____^ `Y` is a good letter too
+  |
+"#]];
+    let renderer = Renderer::plain();
+    assert_data_eq!(renderer.render(input).to_string(), expected);
+}
+#[test]
+fn triple_overlap() {
+    let source = r#"
+fn foo() {
+  X0 Y0 Z0
+  X1 Y1 Z1
+  X2 Y2 Z2
+}
+"#;
+    let input = Level::Error.title("foo").snippet(
+        Snippet::source(source)
+            .line_start(1)
+            .origin("test.rs")
+            .fold(true)
+            .annotation(Level::Error.span(14..38).label("`X` is a good letter"))
+            .annotation(Level::Error.span(17..41).label("`Y` is a good letter too"))
+            .annotation(Level::Error.span(20..44).label("`Z` label")),
+    );
+
+    // This is all `^` until we support primary and secondary labels
+    let expected = str![[r#"
+error: foo
+ --> test.rs:3:3
+  |
+3 |       X0 Y0 Z0
+  |    ___^__^__^
+  |   |___|__|
+  |  ||___|
+  | |||
+4 | |||   X1 Y1 Z1
+5 | |||   X2 Y2 Z2
+  | |||____^__^__^ `Z` label
+  | ||_____|__|
+  | |______|  `Y` is a good letter too
+  |        `X` is a good letter
+  |
+"#]];
+    let renderer = Renderer::plain();
+    assert_data_eq!(renderer.render(input).to_string(), expected);
+}
+#[test]
+fn triple_exact_overlap() {
+    let source = r#"
+fn foo() {
+  X0 Y0 Z0
+  X1 Y1 Z1
+  X2 Y2 Z2
+}
+"#;
+    let input = Level::Error.title("foo").snippet(
+        Snippet::source(source)
+            .line_start(1)
+            .origin("test.rs")
+            .fold(true)
+            .annotation(Level::Error.span(14..38).label("`X` is a good letter"))
+            .annotation(Level::Error.span(14..38).label("`Y` is a good letter too"))
+            .annotation(Level::Error.span(14..38).label("`Z` label")),
+    );
+
+    let expected = str![[r#"
+error: foo
+ --> test.rs:3:3
+  |
+3 | /   X0 Y0 Z0
+4 | |   X1 Y1 Z1
+5 | |   X2 Y2 Z2
+  | |    ^
+  | |____|
+  |      `X` is a good letter
+  |      `Y` is a good letter too
+  |      `Z` label
+  |
+"#]];
+    let renderer = Renderer::plain();
+    assert_data_eq!(renderer.render(input).to_string(), expected);
+}
+#[test]
+fn minimum_depth() {
+    let source = r#"
+fn foo() {
+  X0 Y0 Z0
+  X1 Y1 Z1
+  X2 Y2 Z2
+  X3 Y3 Z3
+}
+"#;
+    let input = Level::Error.title("foo").snippet(
+        Snippet::source(source)
+            .line_start(1)
+            .origin("test.rs")
+            .fold(true)
+            .annotation(Level::Error.span(17..27).label("`X` is a good letter"))
+            .annotation(Level::Error.span(28..44).label("`Y` is a good letter too"))
+            .annotation(Level::Error.span(36..52).label("`Z`")),
+    );
+
+    // This is all `^` until we support primary and secondary labels
+    let expected = str![[r#"
+error: foo
+ --> test.rs:3:6
+  |
+3 |      X0 Y0 Z0
+  |  _______^
+4 | |    X1 Y1 Z1
+  | | ____^_^
+  | ||____|
```

Comment:
This is something I want to handle in rustc eventually: when we have only multiline spans pointing at a line of code, invert the order of the vertical lines so that there are no overlaps anymore. (Future work/Not a blocker)

Reply:
This was added to Rust in rust-lang/rust#126882, and added to this PR in the latest commit